ARROW-9517: [C++/Python] Add support for temporary credentials to S3Options

corleyma opened the pull request:
### Background
AWS provides a mechanism for using temporary credentials to access AWS resources.  When accessing AWS resources with a set of temporary credentials,[ users must provide a session token](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp_use-resources.html) in addition to the usual access key id and secret access key.

This PR adds support for providing a session token when initializing an S3FileSystem.

Additionally, this PR adds support for auto-refreshing temporary credentials via STS AssumeRole: instead of passing explicit credentials, users supply the arn of a role to assume, and an `STSAssumeRoleCredentialsProvider` will be created to handle fetching temporary credentials by assuming this role.

### Changes
#### C++
- updated `S3Options.FromAccessKey` and `S3Options.ConfigureAccessKey` to accept an optional `session_token` argument (defaulting to empty string, in accordance with the convention of [Aws::Auth::AWSCredentials](https://sdk.amazonaws.com/cpp/api/0.12.9/d4/d27/class_aws_1_1_auth_1_1_a_w_s_credentials.html) as implemented in the AWS C++ SDK.)
- updated `S3FileSystem` implementation to initialize S3Client with CredentialsProvider instead of directly passing AwsCredentials (enabling auto-refreshing of AwsCredentials from provider)
- added `S3Options.GetSessionToken` method
- added `S3Options.FromAssumeRole` and `S3Options.ConfigureAssumeRoleCredentials` to support auto-refreshing temporary credentials via STS AssumeRole.

#### Python
- updated cdef `CS3Options` to reflect updates to C++ library
- added optional `session_token` argument to `S3FileSystem.__init__`.

### Testing
- updated `test_s3_options` python unittest to include a mock session_token when initializing S3FileSystem.
- updated `test_s3_options` python unittest to include relevant scenarios for provided a `role_arn` when initializing S3FileSystem.
- updated `s3fs_test.cc` c++ tests to cover changes to FromAccessKey as well as the newly added FromAssumeRole.
- successfully ran ctest suite and pytest suite with `--enable-s3` option.

corleyma commented:
@kszucs Not to be a bother, but I just want to get a sense for whether there is more work to do here or if we can get this merged.  It's an important prerequisite for my team to use Arrow in production.

kszucs commented:
Sorry I didn’t have time because of the release. I’m going to review it tomorrow - and most certainly merge it as well. Thanks for your patience!

corleyma commented:
We're very excited about the 1.0.0 release so no apologies necessary!  

Out of curiosity, what is the likely timeline between merge of this PR and an official release containing these changes?  Is the cadence of one release every few months likely to continue?  Just wondering about whether or not we should be doing our own internal builds.

nealrichardson commented:
> Is the cadence of one release every few months likely to continue?

I'd expect so. Depending on your use case and stability requirements, you could try using our nightly packages.

kszucs commented on cpp/src/arrow/filesystem/s3fs.h:
```diff
@@ -62,10 +62,12 @@ struct ARROW_EXPORT S3Options {
 
   std::string GetAccessKey() const;
   std::string GetSecretKey() const;
+  std::string GetSessionToken() const;
```
Please cover the new API with tests on the C++ side.

corleyma replied:
I added C++ tests to `s3fs_test.cc` to cover these changes.  Unfortunately, adding support for STS-like temporary credentials to Minio involves addition of another dependency, Keycloak.  I spent a bit of time reading relevant docs but could not seem to find a programmatic way to replicate the desired behavior (the Keycloak guides I found were very UI-driven).

pitrou replied:
I think it's ok not to have any deeper test.

pitrou commented on python/pyarrow/_s3fs.pyx:
```diff
@@ -81,9 +81,9 @@ cdef class S3FileSystem(FileSystem):
-    def __init__(self, *, access_key=None, secret_key=None, anonymous=False,
-                 region=None, scheme=None, endpoint_override=None,
-                 bint background_writes=True):
+    def __init__(self, *, access_key=None, secret_key=None, session_token=None,
```
You must add `session_token` in the docstring above.

corleyma replied:
updated docstring accordingly, thanks for catching that.

pitrou commented on python/pyarrow/_s3fs.pyx:
```diff
@@ -105,9 +105,14 @@ cdef class S3FileSystem(FileSystem):
                 tobytes(access_key),
-                tobytes(secret_key)
+                tobytes(secret_key),
+                tobytes(session_token)
```
If `session_token` is only used when either `access_key` or `secret_key` is given, then an error should be raised when `session_token and access_key is None and secret_key is None`.

corleyma replied:
Done.

corleyma commented:
I addressed the comments in the initial round of review, and I also added deeper support for temporary credentials by enabling the use of `STSAssumeRoleCredentialsProvider` to auto-refresh temporary credentials.  However, this seems to break conda-based builds because of an inability to find/link additional dependencies from the aws c++ sdk (sts).  Unfortunately my knowledge of both conda and CMake is a bit limited, so I'm not sure how to go about addressing that.  I'm happy to dig a bit deeper if one of you can point the way.

pitrou commented:
Thank you @corleyma. I'll take a look at the conda-forge part.

pitrou commented on cpp/src/arrow/filesystem/s3fs_test.cc:
```diff
@@ -197,6 +197,32 @@ TEST(S3Options, FromUri) {
+  S3Options options;
+
+  // the only required argument should be role arn
+  options = S3Options::FromAssumeRole("my_role_arn");
```
This test seems to take a very long time on my local machine. Apparently it's trying to make connections to `169.254.169.254` in a loop. Is there a way to avoid that?

corleyma replied:
I added 2 relevant changes:
- I updated the test to set an env var that will cause DefaultAWSCredentialsProviderChain to skip trying the EC2 instance metadata endpoint
- I added the ability to provide your own STSClient instead of using the default configuration, so users can choose to use something other than e.g. the DefaultAWSCredentialsProviderChain.
The test runtime is down from ~5s to ~3ms.

pitrou commented:
I've submitted a PR for conda-forge at https://github.com/conda-forge/aws-sdk-cpp-feedstock/pull/124

corleyma commented on cpp/src/arrow/filesystem/s3fs.h:
```diff
@@ -62,10 +75,18 @@ struct ARROW_EXPORT S3Options {
+  void ConfigureAssumeRoleCredentials(
+      const std::string& role_arn, const std::string& session_name = "",
+      const std::string& external_id = "", int load_frequency = 900,
+      const std::shared_ptr<Aws::STS::STSClient>& stsClient = nullptr);
```
`cpplint` does not like this, but I am just mirroring the default value used by the SDK.  Should I suppress the violation or is there indeed a better practice?  I'm a C++ neophyte so pointers (ha!) appreciated.

corleyma replied:
I inferred this is something to do with compliance with C++/CLI; replaced with existing NULLPTR macro.

corleyma commented:
> I've submitted a PR for conda-forge at [conda-forge/aws-sdk-cpp-feedstock#124](https://github.com/conda-forge/aws-sdk-cpp-feedstock/pull/124)

Thanks for that @pitrou.  Is there anything I need to do to get the CI systems to pick up the new conda-forge packages?

pitrou commented:
>  Is there anything I need to do to get the CI systems to pick up the new conda-forge packages?

I don't think so, perhaps there is a CDN in-between that delays the package updates. Let's try again.

pitrou commented:
@corleyma Hmm, it seems that by force-pushing you deleted the changes I made yesterday.

pitrou commented:
@corleyma Can you apply this patch?

pitrou commented on cpp/src/arrow/filesystem/s3fs_test.cc:
```diff
@@ -197,6 +201,44 @@ TEST(S3Options, FromUri) {
+  // we set this envvar to speed up tests by ensuring
+  // DefaultAWSCredentialsProviderChain does not query (inaccessible)
+  // EC2 metadata endpoint
+  ASSERT_OK(SetEnvVar("AWS_EC2_METADATA_DISABLED", "true"));
```
So, the right way to do this would be to create a test fixture:
```c++
class S3OptionsTest : public ::testing::Test {
 public:
  void SetUp() {
    ASSERT_OK(SetEnvVar("AWS_EC2_METADATA_DISABLED", "true"));
  }
  void TearDown() {
    ASSERT_OK(DelEnvVar("AWS_EC2_METADATA_DISABLED"));
  }
};
```
And then to replace every `TEST(S3Options, ...)` with `TEST_F(S3OptionsTest, ...)`.

corleyma commented:
> @corleyma Can you apply this patch? https://gist.github.com/pitrou/0dc970a2238b9c19b5f8fb991d2fb8f7

Whoops, didn't notice you were playing in this fork too.  I'll quit being destructive and let you guys squash on merge.  Patch applied.

corleyma commented:
@pitrou any idea what the cadence is for rebuilding those Docker images used by the CI system (or whether we can use our own base image)?

corleyma commented:
@github-actions crossbow submit -g conda

pitrou commented:
@corleyma AFAIK, they should have been updated automatically but it looks like they haven't. Will take a look.

corleyma commented:
@github-actions crossbow submit -g conda

corleyma commented:
@pitrou I think whatever you did fixed the Github Actions-based CI; I can't seem to relaunch the ursabot/crossbow-based CI tasks (permissions, perhaps?), but I am guessing if those kick off again everything might be green.

pitrou commented:
@corleyma Unfortunately, the "ursabot" builders use a different execution scheme. The packages need to be updated manually there, I beliave. I pinged @kszucs who should be able to handle it.

kszucs commented:
Indeed, the docker images there need to be rebuilt. I'm updating them.

kszucs commented:
@ursabot build

pitrou commented:
Seems like we're fine now, will merge.
